File: World Finals 2010/A.cpp
// Problem: Letter Stamper
// Language: C++
// Author: KirarinSnow
// Usage: g++ thisfile.cpp -o executable && ./executable <input.in >output.out


#include <iostream>
#include <string>
#include <string.h>

using namespace std;

const int MAX = 1<<20;
int b[7001][6][7001];
char s[7000];
const char *P[6] = {"ABC", "BCA", "CAB", "ACB", "BAC", "CBA"};

int main() {
  int cases;
  cin >> cases;
  
  for (int cas = 1; cas <= cases; cas++) {
    cin >> s;
    int n = strlen(s);

    memset(b, 0, sizeof b);
    for (int p = 0; p < 6; p++) {
      for (int l = 0; l <= n; l++) {
	b[n][p][l] = l;
      }
    }

    for (int k = n-1; k >= 0; k--) {
      char c = s[k];
      for (int p = 0; p < 6; p++) {
	for (int l = 0; l <= n; l++) {
	  if (l == 0) {
	    int v = MAX;
	    for (int pp = 0; pp < 6; pp++) {
	      string sp = P[pp];
	      int d = sp.find(c);
	      if (l+d+1 <= n) {
		int w = 2+d+b[k+1][pp][l+d+1];
		if (w < v) v=w;
	      }
	    }
	    b[k][p][l] = v;
	  }
	  else {
	    int t = P[p][(l-1)%3];
	    if (c == t) {
	      b[k][p][l] = 1+b[k+1][p][l];
	    }
	    else {
	      int v = MAX;
	      if (c == P[p][l%3]) {
		if (l+1 <= n) {
		  int w = 2+b[k+1][p][l+1];
		  if (w < v) v=w;
		}
	      }
	      else {
		if (l+2 <= n) {
		  int w = 3+b[k+1][p][l+2];
		  if (w < v) v=w;
		}
		if (l == 1) {
		  if (l+1 <= n) {
		    int w = 2+b[k+1][(p+3)%6][l+1];
		    if (w < v) v=w;
		  }
		}
	      }
	      if (l >= 2 && c == P[p][(l-2)%3]) {
		int w = 2+b[k+1][p][l-1];
		if (w < v) v=w;
	      }
	      if (l >= 3 && c == P[p][l%3]) {
		int w = 3+b[k+1][p][l-2];
		if (w < v) v=w;
	      }
	      b[k][p][l] = v;
	    }
	  }
	}
      }
    }
    
    int ans = MAX;
    for (int p = 0; p < 6; p++) {
      int w = b[0][p][0];
      if (w < ans) ans = w;
    }

    cout << "Case #" << cas << ": " << ans << endl;
  }
    
  return 0;
}

File: World Finals 2010/A.py
#!/usr/bin/env python
#
# Problem: Letter Stamper
# Language: Python
# Author: KirarinSnow
# Usage: python thisfile.py <input.in >output.out
# Comments: Recursion works only with small.


P = ['ABC', 'BCA', 'CAB', 'ACB', 'BAC', 'CBA']
MAX = 1<<32

def best(s, p, n):
    if (s, p, n) not in b:
        if len(s) == 0:  # empty remainder: pop entire stack
            b[(s, p, n)] = n
        else:
            c = s[0]
            ns = s[1:]
            if n == 0:  # empty stack: push until printable, then print
                v = MAX
                for pp in range(6):
                    d = P[pp].find(c)
                    cost = 2+d+best(ns, pp, n+d+1)
                    v = min(v, cost)
                b[(s, p, n)] = v
            else:
                t = P[p][(n-1)%3]
                if c == t:  # top of remainder == top of stack: print
                    b[(s, p, n)] = 1+best(ns, p, n)
                else:
                    v = MAX
                    if c == P[p][n%3]: # next is OK; can push 1
                        v = min(v, 2+best(ns, p, n+1))
                    else: # next is not OK; can push 2
                        v = min(v, 3+best(ns, p, n+2))
                        if n == 1:  # can also switch pattern and push 1
                            v = min(v, 2+best(ns, (p+3)%6, n+1))
                    if n >= 2 and c == P[p][(n-2)%3]:  # can pop 1
                        v = min(v, 2+best(ns, p, n-1))
                    if n >= 3 and c == P[p][n%3]:  # can pop 2
                        v = min(v, 3+best(ns, p, n-2))
                    b[(s, p, n)] = v
    return b[(s, p, n)]

        
for case in range(int(raw_input())):
    s = raw_input()
    b = {}
    ans = min([best(s, p, 0) for p in range(6)])
    print "Case #%d: %s" % (case+1, ans)
